src/psyclone/psyad/tl2ad.py

```diff
@@ -98,9 +155,407 @@ def generate_adjoint(tl_psyir):
+                            "reference to '{2}' in its shape. However, '{2}' "
+                            "is not passed as an argument. This is not "
+                            "supported.".format(
+                                arg.name, tl_kernel.name, dim.symbol.name))
+                elif isinstance(dim, Literal):
+                    new_shape.append(dim.copy())
+                else:
+                    raise InternalError(
+                        "Argument '{0}' to kernel '{1}' contains a '{2}' in "
+                        "its shape definition but expected an ArrayType."
+                        "Extent, a Reference or a Literal".format(
+                            arg.name, tl_kernel.name, type(dim).__name__))
+            new_sym = symbol_table.new_symbol(arg.name, symbol_type=DataSymbol,
+                                              datatype=ArrayType(arg.datatype,
+                                                                 new_shape))
+        new_arg_list.append(new_sym)
+        # Create variables to hold a copy of the inputs
+        input_sym = symbol_table.new_symbol(new_sym.name+"_input",
+                                            symbol_type=type(new_sym),
+                                            datatype=new_sym.datatype)
+        inputs.append(new_sym)
+        input_copies.append(input_sym)
+
+    statements = []
+    # Initialise those variables and keep a copy of them.
+    # TODO #1247 we need to add comments to the generated code!
+    for sym, sym_copy in zip(inputs, input_copies):
+        # The PSyIR doesn't support the random_number Fortran intrinsic so we
+        # create a CodeBlock for it. Happily, the intrinsic will initialise
+        # all elements of an array passed to it so we don't have to take any
+        # special action.
+        ptree = Fortran2003.Call_Stmt(
+            "call random_number({0})".format(sym.name))
+        statements.append(CodeBlock([ptree], CodeBlock.Structure.STATEMENT))
+        statements.append(
+            Assignment.create(Reference(sym_copy), Reference(sym)))
+
+    # Call the TL kernel
+    statements.append(Call.create(tl_kernel_sym,
+                                  [Reference(sym) for sym in new_arg_list]))
+
+    # Compute the inner product of the result of the TL kernel
+    statements += _create_inner_product(inner1,
+                                        [(sym, sym) for sym in inputs])
+
+    # Call the adjoint kernel using the outputs of the TL kernel as input
+    statements.append(Call.create(adj_kernel_sym,
+                                  [Reference(sym) for sym in new_arg_list]))
+
+    # Compute inner product of result of adjoint kernel with original inputs
+    statements += _create_inner_product(inner2,
+                                        zip(inputs, input_copies))
+
+    # Compare the inner products
+    tol_zero = Literal("1.0e-10", REAL_DOUBLE_TYPE)
```

Comment:
So this should also be in a config file presumably.

Reply:
Yes it should. I've created an issue and also moved the value itself up to a module-wide constant at the top of the file.